### PR TITLE
fix: replace `ArrayBroadcast` in `where` pass

### PR DESCRIPTION
**Pull request by kmr-srbh**
```fortran
program main
   call where_03

contains

   subroutine where_03
      implicit none
      real a(4)

      a = [1.0, 2.0, 1.0, 4.0]

      where(a == 1.0) a = a * 3.0

   end subroutine
end program
```

is transformed to

```fortran
! Fortran code after applying the pass: simplifier
program main
implicit none
call where_03()

contains

subroutine where_03()
    real(4), dimension(4) :: __libasr_created__array_broadcast_
    real(4), dimension(4) :: a
    a = [1.00000000e+00, 2.00000000e+00, 3.00000000e+00, 4.00000000e+00]
    where (a == 1.00000000e+00)
        __libasr_created__array_broadcast_ = 3.00000000e+00
        a = a*__libasr_created__array_broadcast_
    end where
    print *, a
end subroutine where_03

end program main
```

### CODE REVIEW

**Comment by kmr-srbh** [comment]
This is later transformed to the following in the `where` pass:

```fortran
! Fortran code after applying the pass: where
program main
implicit none
call where_03()

contains

subroutine where_03()
    integer(4) :: __1_k
    real(4), dimension(4) :: __libasr_created__array_broadcast_
    real(4), dimension(4) :: a
    a = [1.00000000e+00, 2.00000000e+00, 1.00000000e+00, 4.00000000e+00]
    do __1_k = lbound(a, 1), ubound(a, 1), 1
        if (a(__1_k) == 1.00000000e+00) then
            __libasr_created__array_broadcast_(__1_k) = 3.00000000e+00
            a(__1_k) = a(__1_k)*__libasr_created__array_broadcast_(__1_k)
        end if
    end do
end subroutine where_03

end program main
```

**Comment by kmr-srbh** [comment]
The test `expr_17` fails with this change. Please let me know if the change is correct, then we can handle this new failure as a separate issue.

File: src/libasr/pass/simplifier.cpp [hunk 1]
```diff
@@ -1541,6 +1541,7 @@ class ReplaceExprWithTemporary: public ASR::BaseExprReplacer<ReplaceExprWithTemp
             replace_expr(x->m_value);
             current_expr = current_expr_copy_163;
         }
+        replace_current_expr("_array_broadcast_")
```
Comment (gxyd) [inline src/libasr/pass/simplifier.cpp:1544]: I am a little doubtful if this is actually the right fix for it, maybe it masks the real problem we've here, another reason being that whichever ASR node needs a temporary variable for most we already listed it in ASR.asdl https://github.com/lfortran/lfortran/blob/simplifier_pass/src/libasr/ASR.asdl#L90, for which needs a temporary variable, and ArrayBroadcast https://github.com/lfortran/lfortran/blob/simplifier_pass/src/libasr/ASR.asdl#L151 probably doesn't need a temporary variable.

Comment (gxyd) [reply src/libasr/pass/simplifier.cpp:1544]: Also, I've never worked with `where` pass myself, and I wouldn't know the details of it, I remember reading somewhere that `where` pass is currently a bit hackish and not clean. (See: https://github.com/lfortran/lfortran/pull/4406#issuecomment-2198453478)

Maybe @Pranavchiku has some idea whether this looks like good fix?

**Comment by Pranavchiku** [comment]
```diff
diff --git a/a.f90 b/a.f90
index b3abfe473..a565ac3d5 100644
--- a/a.f90
+++ b/a.f90
@@ -9,8 +9,8 @@ subroutine where_03()
     real(4), dimension(4) :: __libasr_created__array_broadcast_
     real(4), dimension(4) :: a
     a = [1.00000000e+00, 2.00000000e+00, 3.00000000e+00, 4.00000000e+00]
+    __libasr_created__array_broadcast_ = 3.00000000e+00
     where (a == 1.00000000e+00)
-        __libasr_created__array_broadcast_ = 3.00000000e+00
         a = a*__libasr_created__array_broadcast_
     end where
     print *, a
```

Either this or original one is fine I believe.

**Review by Pranavchiku — COMMENTED** [review]
The approach is good, do make progress in same direction

**Comment by kmr-srbh** [comment]
Ready.

**Comment by Pranavchiku** [comment]
Can you please post a full transformation in steps after simplifier and where pass.

**Comment by kmr-srbh** [comment]
After `simplifier` (no change):

```fortran
! Fortran code after applying the pass: simplifier
program main
implicit none
call where_03()

contains

subroutine where_03()
    real(4), dimension(4) :: a
    a = [1.00000000e+00, 2.00000000e+00, 1.00000000e+00, 4.00000000e+00]
    where (a == 1.00000000e+00)
        a = a*3.00000000e+00
    end where
end subroutine where_03

end program main
```

After `where` (we internally set the `current_expr` from `ArrayBroadcast` to `x.m_array`):

<details>
```clojure
;; ASR after applying the pass: where
(TranslationUnit
    (SymbolTable
        1
        {
            main:
                (Program
                    (SymbolTable
                        2
                        {
                            where_03:
                                (Function
                                    (SymbolTable
                                        3
                                        {
                                            __1_k:
                                                (Variable
                                                    3
                                                    __1_k
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Integer 4)
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                ),
                                            a:
                                                (Variable
                                                    3
                                                    a
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Array
                                                        (Real 4)
                                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                                        (IntegerConstant 4 (Integer 4) Decimal))]
                                                        FixedSizeArray
                                                    )
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                )
                                        })
                                    where_03
                                    (FunctionType
                                        []
                                        ()
                                        Source
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        .false.
                                    )
                                    []
                                    []
                                    [(Assignment
                                        (Var 3 a)
                                        (ArrayConstant
                                            16
                                            [1.00000000e+00, 2.00000000e+00, 1.00000000e+00, 4.00000000e+00]
                                            (Array
                                                (Real 4)
                                                [((IntegerConstant 1 (Integer 4) Decimal)
                                                (IntegerConstant 4 (Integer 4) Decimal))]
                                                FixedSizeArray
                                            )
                                            ColMajor
                                        )
                                        ()
                                    )
                                    (DoLoop
                                        ()
                                        ((Var 3 __1_k)
                                        (ArrayBound
                                            (Var 3 a)
                                            (IntegerConstant 1 (Integer 4) Decimal)
                                            (Integer 4)
                                            LBound
                                            ()
                                        )
                                        (ArrayBound
                                            (Var 3 a)
                                            (IntegerConstant 1 (Integer 4) Decimal)
                                            (Integer 4)
                                            UBound
                                            ()
                                        )
                                        (IntegerConstant 1 (Integer 4) Decimal))
                                        [(If
                                            (RealCompare
                                                (ArrayItem
                                                    (Var 3 a)
                                                    [(()
                                                    (Var 3 __1_k)
                                                    ())]
                                                    (Real 4)
                                                    RowMajor
                                                    ()
                                                )
                                                Eq
                                                (RealConstant
                                                    1.000000
                                                    (Real 4)
                                                )
                                                (Logical 4)
                                                ()
                                            )
                                            [(Assignment
                                                (ArrayItem
                                                    (Var 3 a)
                                                    [(()
                                                    (Var 3 __1_k)
                                                    ())]
                                                    (Real 4)
                                                    RowMajor
                                                    ()
                                                )
                                                (RealBinOp
                                                    (ArrayItem
                                                        (Var 3 a)
                                                        [(()
                                                        (Var 3 __1_k)
                                                        ())]
                                                        (Real 4)
                                                        RowMajor
                                                        ()
                                                    )
                                                    Mul
                                                    (RealConstant
                                                        3.000000
                                                        (Real 4)
                                                    )
                                                    (Array
                                                        (Real 4)
                                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                                        (IntegerConstant 4 (Integer 4) Decimal))]
                                                        FixedSizeArray
                                                    )
                                                    ()
                                                )
                                                ()
                                            )]
                                            []
                                        )]
                                        []
                                    )]
                                    ()
                                    Public
                                    .false.
                                    .false.
                                    ()
                                )
                        })
                    main
                    []
                    [(SubroutineCall
                        2 where_03
                        ()
                        []
                        ()
                    )]
                )
        })
    []
)
```
</details>

**Review by certik — APPROVED** [review]

**Pull request merged**